pkg/controller/servicemesh/controlplane/reconciler.go

```diff
@@ -208,20 +231,116 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
+// base values, giving preference to the base values for conflicts
+func mergeValues(base map[string]interface{}, input map[string]interface{}) map[string]interface{} {
+	if base == nil {
+		base = make(map[string]interface{}, 1)
+	}
+
+	for key, value := range input {
+		//if the key doesn't already exist, add it
+		if _, exists := base[key]; !exists {
+			base[key] = value
+			continue
+		}
+
+		// at this point, key exists in both input and base.
+		// If both are maps, recurse.
+		// If only input is a map, ignore it. We don't want to overrwrite base.
+		// If both are values, again, ignore it since we don't want to overrwrite base.
+		if baseKeyAsMap, baseOK := base[key].(map[string]interface{}); baseOK {
+			if inputAsMap, inputOK := value.(map[string]interface{}); inputOK {
+				base[key] = mergeValues(baseKeyAsMap, inputAsMap)
+			}
+		}
+	}
+	return base
+}
+
+func (r *ControlPlaneReconciler) getSMCPTemplate(name string) (v1.ControlPlaneSpec, error) {
+	if strings.Contains(name, "/") {
+		return v1.ControlPlaneSpec{}, fmt.Errorf("template name contains invalid character '/'")
+	}
+
+	templateContent, err := ioutil.ReadFile(smcpTemplatePath + name)
+	if err != nil {
+		//if we can't read from the user template path, try from the default path
+		//we use two paths because Kubernetes will not auto-update volume mounted
+		//configmaps mounted in directories with pre-existing content
+		defaultTemplateContent, defaultErr := ioutil.ReadFile(smcpDefaultTemplatePath + name)
+		if defaultErr != nil {
+			return v1.ControlPlaneSpec{}, fmt.Errorf("template cannot be loaded from user or default directory. Error from user: %s. Error from default: %s", err, defaultErr)
+		}
+		templateContent = defaultTemplateContent
+	}
+
+	var template v1.ServiceMeshControlPlane
+	if err = yaml.Unmarshal(templateContent, &template); err != nil {
+		return v1.ControlPlaneSpec{}, fmt.Errorf("failed to parse template %s contents: %s", name, err)
+
+	}
+	return template.Spec, nil
+}
+
+func (r *ControlPlaneReconciler) renderSMCPTemplates(smcp v1.ControlPlaneSpec, visited map[string]struct{}) (v1.ControlPlaneSpec, error) {
+	if smcp.Template == "" {
+		return smcp, nil
+	}
+	r.Log.Info(fmt.Sprintf("processing smcp template %s", smcp.Template))
+
+	if _, ok := visited[smcp.Template]; ok {
+		return smcp, fmt.Errorf("SMCP templates form cyclic dependency. Cannot proceed")
+	}
+
+	template, err := r.getSMCPTemplate(smcp.Template)
+	if err != nil {
+		return smcp, err
+	}
+
+	template, err = r.renderSMCPTemplates(template, visited)
+	if err != nil {
+		r.Log.Info(fmt.Sprintf("error rendering SMCP templates: %s\n", err))
+		return smcp, err
+	}
+
+	visited[smcp.Template] = struct{}{}
+
+	smcp.Istio = mergeValues(smcp.Istio, template.Istio)
+	smcp.ThreeScale = mergeValues(smcp.ThreeScale, template.ThreeScale)
+	return smcp, nil
+}
+
+func (r *ControlPlaneReconciler) updateSMCPWithTemplates(smcpSpec v1.ControlPlaneSpec) (v1.ControlPlaneSpec, error) {
```

Comment:
nit, naming.  maybe `applyTemplates()`.  conflicts a bit with `applyTemplate()` above, but it seems like this is applying the templates to the smcp resource, while the above is recursive, going through each template referenced by the template.

Reply:
Correct. This one is "applying" the template, or updating the SMCP with the associated template values. Whereas the other one traverses all templates before processing them.

Reply:
Done